gapis/api/transform2/transform_chain.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/commandGenerator"
+	"github.com/google/gapid/gapis/config"
+)
+
+type transformChain struct {
+	transforms       []Transform
+	out              Writer
+	generator        commandGenerator.CommandGenerator
+	hasBegun         bool
+	hasEnded         bool
+	currentCommandID api.CmdID
+}
+
+// CreateTransformChain creates a transform chain that will run
+// the required transforms on every command coming from commandGenerator
+func CreateTransformChain(generator commandGenerator.CommandGenerator, transforms []Transform, out Writer) *transformChain {
+	return &transformChain{
+		generator:        generator,
+		transforms:       transforms,
+		out:              out,
+		hasBegun:         false,
+		hasEnded:         false,
+		currentCommandID: 0,
+	}
+}
+
+func (chain *transformChain) beginChain(ctx context.Context) error {
+	handleInitialState(chain.out.State())
+
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.BeginTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "Begin Transform Error [%v] : %v", transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
```

Comment:
How do we know this is updating the correct state? It looks like it's applying the mutation to the final state?